Remove UploadButtonElement

annevk opened the pull request:
#### 1d24a443395cd438712eee30e17e527395f37dd0
<pre>
Remove UploadButtonElement
<a href="https://bugs.webkit.org/show_bug.cgi?id=117777">https://bugs.webkit.org/show_bug.cgi?id=117777</a>

Reviewed by Tim Nguyen.

A custom PseudoId does not require a subclass.

Based on:
<a href="https://chromium.googlesource.com/chromium/blink/+/0ef88d710ad04158caa9835932776dbe1d40f505">https://chromium.googlesource.com/chromium/blink/+/0ef88d710ad04158caa9835932776dbe1d40f505</a>

Also refactor getting a reference to the button into a static method.

* Source/WebCore/html/FileInputType.cpp:
(WebCore::FileInputType::createShadowSubtree):
(WebCore::FileInputType::disabledStateChanged):
(WebCore::FileInputType::attributeChanged):
(isType): Deleted.
(): Deleted.
(WebCore::UploadButtonElement::create): Deleted.
(WebCore::UploadButtonElement::createForMultiple): Deleted.
(WebCore::UploadButtonElement::createInternal): Deleted.
(WebCore::UploadButtonElement::UploadButtonElement): Deleted.

Canonical link: <a href="https://commits.webkit.org/272539@main">https://commits.webkit.org/272539@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/118d460cb643ccf514cfd9aa29fb70693910526a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32007 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10711 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33791 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34531 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29005 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32819 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13073 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7945 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28604 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32387 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9066 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28623 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7856 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8032 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28538 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35875 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29128 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28992 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34134 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8121 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6103 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31992 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9765 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7481 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8775 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8644 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/FileInputType.cpp:
```diff
@@ -276,7 +224,11 @@ void FileInputType::createShadowSubtree()
 
-    auto button = element()->multiple() ? UploadButtonElement::createForMultiple(element()->document()) : UploadButtonElement::create(element()->document());
+    auto button = HTMLInputElement::create(inputTag, element()->document(), nullptr, false);
+    ScriptDisallowedScope::EventAllowedScope eventAllowedScopeBeforeAppend { button };
```
To limit the scope you could either wrap this code inside brackets:
```
    auto button = HTMLInputElement::create(inputTag, element()->document(), nullptr, false);
    {
        ScriptDisallowedScope::EventAllowedScope eventAllowedScopeBeforeAppend { button };
        button->setType(InputTypeNames::button());
        button->setPseudo(ShadowPseudoIds::fileSelectorButton());
        button->setValue(element()->multiple() ? fileButtonChooseMultipleFilesLabel() : fileButtonChooseFileLabel());
    }
 ```

or inside a lambda. Otherwise the scope exception will be extended until the end of the method.

nt1m commented on Source/WebCore/html/FileInputType.cpp:
```diff
@@ -289,8 +241,8 @@ void FileInputType::disabledStateChanged()
-    
-    if (RefPtr button = childrenOfType<UploadButtonElement>(*root).first())
+
+    if (RefPtr button = childrenOfType<HTMLInputElement>(*root).first())
```
Can we make this less fragile? Repeating this logic isn't great. There are 2 ways to do this:
- create a helper method (`FileInputType::uploadButtonElement()`?) that returns `childrenOfType<HTMLInputElement>(*root).first()`
- store the `uploadButton` as a member on `FileInputType`

You'll also be able to de-duplicate the logic that gets the `userAgentShadowRoot`and the `element`.

nt1m commented on Source/WebCore/html/FileInputType.cpp:
```diff
@@ -276,32 +224,43 @@ void FileInputType::createShadowSubtree()
     disabledStateChanged();
 }
 
+static RefPtr<HTMLInputElement> fileSelectorButton(RefPtr<Element> element)
```
```suggestion
static RefPtr<HTMLInputElement> fileSelectorButton(const Element& element)
```

Instead of asserting the presence of element, you could just take in a reference. (Also you shouldn't pass in a smart pointer here, the caller is supposed to protect the element in this case and we're also just calling trivial getters on element)

annevk committed:
Remove UploadButtonElement
https://bugs.webkit.org/show_bug.cgi?id=117777

Reviewed by Tim Nguyen.

A custom PseudoId does not require a subclass.

Based on:
https://chromium.googlesource.com/chromium/blink/+/0ef88d710ad04158caa9835932776dbe1d40f505

Also refactor getting a reference to the button into a static method.

* Source/WebCore/html/FileInputType.cpp:
(WebCore::FileInputType::createShadowSubtree):
(WebCore::FileInputType::disabledStateChanged):
(WebCore::FileInputType::attributeChanged):
(isType): Deleted.
(): Deleted.
(WebCore::UploadButtonElement::create): Deleted.
(WebCore::UploadButtonElement::createForMultiple): Deleted.
(WebCore::UploadButtonElement::createInternal): Deleted.
(WebCore::UploadButtonElement::UploadButtonElement): Deleted.

Canonical link: https://commits.webkit.org/272539@main
